[Type Checker] Refactor duplicated code

echebbi opened the pull request:
Remove duplicated code in ALE's type checker to make the code easier to maintain.

## Changes

 - introduce a new `TypeChecker` class responsible of checking compatibility between types
 - refactor some methods of `TypeValidator` to remove duplicated code
 - some error messages have new text and position to make them closer to what is usually done (e.g. by JDT); see changes in `TypeValidatorTest`

## New features

 - check type of variables when using the `+=` and `-=` operators (until now [a warning was shown](https://github.com/gemoc/ale-lang/pull/69))
 - check the type of variables declared within _for each_ loops
 - properly retrieve generic type parameters for methods return type (and thus for the `result` variable) and attribute declaration
 - check types when assigning a value to `result` (`:=`, `+=` and `-=` operators)

dvojtise commented:
looks good to me  (anyway, it need to be rebased and checked before being accepted)

echebbi commented:
I implemented missing tests and rebased the branch onto master. CI looks happy so I think this branch can be merged as soon as the review approves it.
